torch/distributed/fsdp/_runtime_utils.py

```diff
@@ -928,25 +930,27 @@ def _register_post_forward_hooks(
 
 
 @no_type_check
-def _register_root_pre_forward_hooks(
+def _register_root_pre_forward_hook(
     state: _FSDPState,
-    modules: Iterable[nn.Module],
+    module: nn.Module,
 ):
     """
-    # TODO (awgu): This requires kwarg support for hooks registered by
-    ``register_forward_pre_hook()``. ``_root_pre_forward()`` does not have the
-    supported hook signature right now.
+    Registers root pre-forward hook on ``module``, which should be the local
+    FSDP root.
+
+    NOTE: For the current composable FSDP design, we have each application of
+    ``fully_shard()`` to a module to indicate that that module is the local
+    FSDP root. We may remove this assumption in the future, in which case we
+    will need to register this root pre-forward hook on any candidate module
```

Comment:
curious, are you referring to after having a dedicated API marking the global root, `fully_shard` might no longer mark the FSDP root? Or is this referring to sth else?

Reply:
For context, let me recap the manual wrapping approach using the `FullyShardedDataParallel` wrapper:
- The user may wrap submodules with `FullyShardedDataParallel`, each following some eager initialization semantics (like constructing and sharding the instance's `FlatParameter`).
- Then, the first forward triggers some lazy initialization semantics (via `_lazy_init()`). Each `FullyShardedDataParallel` instance runs `_lazy_init()` in its pre-forward logic, and the first to call it is marked as the root instance (taking advantage of the module tree structure).

We may imagine a similar approach for `fully_shard()`.
- The user may apply `fully_shard()` to submodules, each following some eager initialization semantics.
- Then, the first forward triggers some lazy initialization semantics (via `_lazy_init()`). We would need to register `_root_pre_forward()` on the module passed to `fully_shard()` _for each_ application of `fully_shard()`. This is what the comment is referring to.

The current design assumes that each application of `fully_shard()` yields a standalone FSDP subtree without any data structure sharing across applications of `fully_shard()`. IIUC, this suffices for H2 2022.

A new design may want to share data structures across applications of `fully_shard()`. If we pursue such a design, we will encounter the case described above that is similar to manual wrapping.

Note that the multiple applications of `fully_shard()` differs from manual wrapping using `FullyShardedDataParallel` in that each application of `fully_shard()` can construct multiple `FlatParameter`s, not just one, since each can represent a logical auto wrapping.